### PR TITLE
fix: codex-doctor stale-state recovery (#171) + clear 8 Dependabot advisories

### LINKED ISSUE #171: Auth token invalidated error does not fail over to next healthy account
## Bug Description

When one stored OAuth account returns:

```text
Your authentication token has been invalidated. Please try signing in again. (run `opencode auth login` if this persists)
```

`oc-codex-multi-auth` appears to keep selecting that account for the requested model family instead of cooling it down/removing it or failing over to another valid stored account.

In this case, re-running `opencode auth login` did not fix the normal plugin path because the persisted family routing still pointed at the bad account slot. Manually switching `activeIndex` and every `activeIndexByFamily` entry from account slot `0` to account slot `1` immediately restored normal operation.

## Steps to Reproduce

1. Configure `oc-codex-multi-auth` with two OpenAI OAuth accounts.
2. Have account slot `0` selected in `~/.opencode/oc-codex-multi-auth-accounts.json`:

```json
{
  "activeIndex": 0,
  "activeIndexByFamily": {
    "gpt-5.4-mini": 0,
    "gpt-5.4": 0,
    "codex": 0
  }
}
```

3. Account slot `0` starts returning the OpenAI/Codex auth error:

```text
Your authentication token has been invalidated. Please try signing in again. (run `opencode auth login` if this persists)
```

4. Run a normal request through the plugin:

```bash
opencode run -m openai/gpt-5.4-mini "Reply exactly: OK"
```

5. Compare with `--pure`, which bypasses the plugin:

```bash
opencode --pure run -m openai/gpt-5.4-mini "Reply exactly: OK"
opencode --pure run -m openai/gpt-5.5-fast "Reply exactly: OK"
```

## Expected Behavior

When a selected account returns this token-invalidated auth error, the plugin should treat it as an auth/account health failure and either:

- refresh/re-authenticate that account if possible,
- mark the account or refresh-token group as cooling down/unhealthy,
- remove it after the configured auth failure threshold,
- or at least fail over to the next healthy account in the account pool for that request/model family.

`codex-health` / `codex-doctor --fix` should also be able to surface or repair this state without requiring manual JSON edits.

## Actual Behavior

Normal plugin mode failed repeatedly with:

```text
Error: Your authentication token has been invalidated. Please try signing in again. (run `opencode auth login` if this persists)
```

The failure persisted after re-login because the plugin continued routing the model family to account slot `0`.

At the same time, bypassing the plugin worked:

```text
opencode --pure run -m openai/gpt-5.4-mini "Reply exactly: OK"
# OK

opencode --pure run -m openai/gpt-5.5-fast "Reply exactly: OK"
# OK
```

After manually changing the multi-account routing to account slot `1`, normal plugin mode worked again:

```text
opencode run -m openai/gpt-5.4-mini "Reply exactly: OK"
# OK

opencode run -m openai/gpt-5.5-fast "Reply exactly: OK"
# OK
```

## Environment

- opencode version: `1.17.3`
- oc-codex-multi-auth version: `6.3.2`
- OS: macOS 26.2
- Node.js: `v26.0.0`
- Account pool: 2 OpenAI OAuth accounts

## Notes / Possible Cause

This looks related to the request/stream error path rather than the token refresh path:

- the plugin already has auth refresh failure handling, cooldown, and removal thresholds;
- rate limits, server errors, unsupported models, and deactivated workspaces appear to have rotation paths;
- but this token-invalidated `AI_APICallError` appears to bubble out as a normal request failure and does not mark the selected account unhealthy or try the next account.

Related but not identical to #156. That issue covered duplicate accounts / disabled valid logins. This case had two stored accounts and was fixed by moving family routing from the invalidated slot to the healthy slot.

## Compliance Checklist

- [x] I'm using this plugin for personal development only
- [x] I have an active ChatGPT Plus/Pro subscription
- [x] This issue is not related to attempting commercial use or TOS violations
- [x] I've reviewed the FAQ and Troubleshooting sections

### PR DESCRIPTION
**Pull request by ndycode**
## Summary

Two related follow-ups to #171 / #172, plus the open Dependabot advisories, in one PR.

### 1. Recovery fix (#171)

After the 401 failover fix shipped in v6.3.3, the reporter documented (issue #171, comment 4748562388) that the plugin could still leave **every account ineligible** even though the credentials were alive — `opencode --pure` and host OAuth both worked, but normal plugin requests failed until the account file was hand-edited.

Reproducing the reporter's exact on-disk state through the real `persistAccountPool` / `loadAccounts` / doctor paths isolated the root cause: it is **not** the login path. The login dedup already merges correctly when the re-login resolves the org identity or shares an email. What was missing is a **recovery path for stale account-health state** that nothing in the normal request flow clears:

- an `auth-failure` / `network-error` `coolingDownUntil` left over from a transient failure that has since recovered, and/or
- `rateLimitResetTimes` markers pointing at a future reset that no longer reflects real quota.

`codex-doctor --fix` refreshed tokens and tried to switch to the healthiest account, but left that stale state in place — so the auto-switch found **no eligible account** and the dead routing persisted across restarts.

**Fix:**
- **`lib/accounts/stale-state.ts` (new, pure/unit-tested):** `clearRefreshedAccountsStaleState` removes cooldown + `rateLimitResetTimes` from accounts whose refresh succeeded this run; `findDisabledTokenSourceDuplicates` detects a disabled `accountIdSource:"token"` entry shadowing an enabled org-backed account by email.
- **`codex-doctor --fix`** now clears stale cooldown/rate-limit state on refreshed accounts (so the existing auto-switch finds an eligible account), and clears the stale TUI quota cache so diagnostics stop referencing an account index/count that no longer matches the pool.
- **`codex-doctor`** surfaces a finding for disabled token-source duplicates with a `codex-remove` hint. **Flagged, never auto-removed** — the only link between a token-source and an org-source record is email, and email-only merges must not collapse distinct multi-org accounts (#64).

This intentionally does **not** change login-time dedup (would risk regressing #64 / #85 for low marginal value). The separate `Error: Aborted` symptom from comment 4748602475 is a different code path and is not addressed here.

### 2. Dependency security (clears all 8 Dependabot advisories)

`npm audit` goes from 7 reported vulnerabilities to **0**. No major version bumps — overrides/patch-level only, matching the repo's existing override pattern (e.g. the esbuild pin).

| Package | From → To | Severity | How |
|---|---|---|---|
| `hono` | 4.12.23 → 4.12.26 | 1 high + 4 moderate | direct dep + override; also clears the transitive `@openauthjs/openauth` alert (via hono) |
| `vite` | ^7.3.2 → ^7.3.5 | 1 high + 1 moderate | override, dev-only via vitest |
| `@babel/core` | 7.28.0 → ^7.29.6 | 1 low | new override, transitive via `@opentui/solid`, stays on 7.x |
| `brace-expansion` 5.x | 5.0.5 → ^5.0.6 | moderate | new override range (existing one only covered 2.x) |

GHSAs: hono GHSA-88fw-hqm2-52qc (high) + GHSA-j6c9-x7qj-28xf / GHSA-rv63-4mwf-qqc2 / GHSA-wgpf-jwqj-8h8p / GHSA-wwfh-h76j-fc44; vite GHSA-fx2h-pf6j-xcff (high) + GHSA-v6wh-96g9-6wx3; @babel/core GHSA-4x5r-pxfx-6jf8.

## Tests / verification
- `test/stale-state.test.ts` — 11 unit tests for both helpers.
- `test/index.test.ts` — 2 end-to-end `codex-doctor` tests (stale-state recovery reproduces the reporter's exact state; duplicate finding).
- Full suite: **2514 passed**, 1 skipped. `tsc --noEmit`, `eslint`, and `npm run build` clean. `npm audit`: **0 vulnerabilities**.

Fixes #171. Supersedes #174 (dependency fixes folded in here).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**note:** greptile review for oc-chatgpt-multi-auth. cite files like `lib/foo.ts:123`. confirm regression tests + windows concurrency/token redaction coverage.
---

<h3>Greptile Summary</h3>

adds a `codex-doctor --fix` stale-state recovery path (issue #171): after a successful token refresh, stale `coolingDownUntil` and `rateLimitResetTimes` fields are now cleared and written atomically so the account re-enters rotation, and disabled `accountIdSource:"token"` duplicates shadowing org-backed accounts are surfaced as a warning with a `codex-remove` hint. seven dependabot advisories are also cleared via hono/vite/babel/brace-expansion bumps.

- **`lib/accounts/stale-state.ts` (new):** pure helpers `clearRefreshedAccountStaleState` / `clearRefreshedAccountsStaleState` mutate account objects in place and return what changed; `findDisabledTokenSourceDuplicates` detects disabled token-source entries shadowing enabled org accounts by email.
- **`lib/tools/codex-doctor.ts`:** stale-state cleared and TUI quota cache purged only when `changedByRefresh` is true; all `appliedFixes` messages are pushed inside the `saveAccounts` try block so a disk-write failure never prints a phantom "Cleared …" message; both code paths are covered by three new end-to-end tests including a `saveAccounts`-throws case.
- **`package.json` / `package-lock.json`:** hono → 4.12.26, vite override → ^7.3.5, new `@babel/core` and `brace-expansion@>=5.0.0 <5.0.6` overrides — patch-level bumps matching the existing override pattern, `npm audit` now reports 0 vulnerabilities.

<h3>Confidence Score: 5/5</h3>

safe to merge — the recovery logic is well-scoped, messages are correctly gated on a successful disk write, and the dependency bumps are patch-level with no API surface changes.

the stale-state helpers are pure functions with 11 dedicated unit tests; the doctor integration adds the saveAccounts-throws case that was previously flagged as missing; the redundant changedByRefresh re-assignment and the missing EBUSY test are informational nits that don't affect routing correctness or data safety.

lib/tools/codex-doctor.ts — the changedByRefresh re-assignment after the stale-summary block is worth cleaning up before this pattern is extended further.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lib/accounts/stale-state.ts | new pure-function module: clearRefreshedAccountStaleState, clearRefreshedAccountsStaleState, findDisabledTokenSourceDuplicates — all well-defined, side-effect-free interfaces with solid unit coverage |
| lib/tools/codex-doctor.ts | integrates stale-state helpers and TUI cache clear into --fix path; messages correctly gated on saveAccounts success; minor redundant changedByRefresh re-assignment when stale state clears but a token refresh already set the flag |
| test/stale-state.test.ts | 11 unit tests covering cooldown clear, rate-limit clear, expired cooldown, clean account no-op, multi-account aggregation, and all four duplicate-detection edge cases including case-insensitive email matching |
| test/index.test.ts | adds three e2e codex-doctor tests: stale-state recovery happy path, disabled token-source duplicate finding, and saveAccounts failure ensuring 'Cleared …' messages are not emitted when persist fails |
| package.json | hono bumped to 4.12.26, vite overridden to ^7.3.5, new @babel/core and brace-expansion@5.x overrides added — matches existing override pattern, clears all 8 Dependabot advisories |
| CHANGELOG.md | unreleased section updated with Fixed and Security entries accurately describing both recovery changes and all four dependency bumps with GHSA references |

</details>

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[codex-doctor --fix] --> B[loadAccounts]
    B --> C{accounts exist?}
    C -- no --> Z[skip fix block]
    C -- yes --> D[queuedRefresh loop]
    D --> E{refresh type === success?}
    E -- yes --> F[update tokens in memory\nrefreshedAccounts.push]
    E -- no/throw --> G[fixErrors.push]
    F --> H{refreshedAccounts.length > 0?}
    H -- yes --> I[clearRefreshedAccountsStaleState\ncooldown + rateLimitResetTimes]
    I --> K{changedByRefresh?}
    H -- no --> K
    K -- yes --> L[saveAccounts]
    L -- success --> M[appliedFixes: Refreshed N tokens\nCleared cooldown\nCleared rate-limit markers]
    L -- throws --> N[fixErrors: Failed to persist]
    K -- no --> O[skip save]
    M --> P{changedByRefresh?}
    N --> P
    P -- yes --> Q[clearTuiQuotaSnapshot]
    Q -- success --> R[appliedFixes: Cleared TUI cache]
    Q -- EBUSY/throws --> S[fixErrors: Failed to clear TUI cache\nWindows EBUSY possible]
    P -- no --> T[AccountManager.loadFromDisk\nauto-switch evaluation]
    R --> T
    S --> T
    T --> U{eligible account found?}
    U -- yes --> V[saveAccounts activeIndex\nappliedFixes: Switched to N]
    U -- no --> W[appliedFixes: No eligible account]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[codex-doctor --fix] --> B[loadAccounts]
    B --> C{accounts exist?}
    C -- no --> Z[skip fix block]
    C -- yes --> D[queuedRefresh loop]
    D --> E{refresh type === success?}
    E -- yes --> F[update tokens in memory\nrefreshedAccounts.push]
    E -- no/throw --> G[fixErrors.push]
    F --> H{refreshedAccounts.length > 0?}
    H -- yes --> I[clearRefreshedAccountsStaleState\ncooldown + rateLimitResetTimes]
    I --> K{changedByRefresh?}
    H -- no --> K
    K -- yes --> L[saveAccounts]
    L -- success --> M[appliedFixes: Refreshed N tokens\nCleared cooldown\nCleared rate-limit markers]
    L -- throws --> N[fixErrors: Failed to persist]
    K -- no --> O[skip save]
    M --> P{changedByRefresh?}
    N --> P
    P -- yes --> Q[clearTuiQuotaSnapshot]
    Q -- success --> R[appliedFixes: Cleared TUI cache]
    Q -- EBUSY/throws --> S[fixErrors: Failed to clear TUI cache\nWindows EBUSY possible]
    P -- no --> T[AccountManager.loadFromDisk\nauto-switch evaluation]
    R --> T
    S --> T
    T --> U{eligible account found?}
    U -- yes --> V[saveAccounts activeIndex\nappliedFixes: Switched to N]
    U -- no --> W[appliedFixes: No eligible account]
```

</a>

<a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22ndycode%2Foc-codex-multi-auth%22%20on%20the%20existing%20branch%20%22fix%2Fissue-171-recovery-gaps%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22fix%2Fissue-171-recovery-gaps%22.%0A%0AFix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Alib%2Ftools%2Fcodex-doctor.ts%3A155-159%0A**redundant%20%60changedByRefresh%20%3D%20true%60%20in%20stale-summary%20block**%0A%0A%60refreshedAccounts.length%20%3E%200%60%20is%20only%20true%20when%20at%20least%20one%20account%20passed%20the%20%60refreshResult.type%20%3D%3D%3D%20%22success%22%60%20branch%2C%20which%20already%20sets%20%60changedByRefresh%20%3D%20true%60%20inside%20the%20loop.%20the%20second%20assignment%20here%20can%20never%20flip%20the%20flag%20from%20%60false%60%20to%20%60true%60%20%E2%80%94%20it's%20always%20redundant.%20a%20future%20dev%20adding%20a%20path%20where%20stale%20state%20is%20cleared%20*without*%20a%20successful%20token%20refresh%20%28e.g.%20a%20manual%20cooldown%20reset%29%20would%20silently%20rely%20on%20this%20flag%20being%20set%20by%20the%20refresh%20loop%2C%20creating%20a%20subtle%20dependency%20that%20isn't%20obvious%20at%20all.%0A%0A%23%23%23%20Issue%202%20of%202%0Atest%2Fstale-state.test.ts%3A1-10%0A**missing%20vitest%20coverage%3A%20%60clearTuiQuotaSnapshot%60%20EBUSY%20failure%20path**%0A%0Athe%20catch%20block%20at%20%60codex-doctor.ts%3A200-208%60%20explicitly%20calls%20out%20the%20windows%20EBUSY%20scenario%20%28TUI%20holds%20the%20cache%20file%20open%29%2C%20but%20there's%20no%20test%20that%20mocks%20%60clearTuiQuotaSnapshot%60%20to%20throw%20and%20asserts%20the%20error%20lands%20in%20%60fixErrors%60%20rather%20than%20crashing%20the%20tool.%20given%20this%20codebase%20ships%20on%20windows%2C%20a%20test%20locking%20in%20that%20error-surface%20behaviour%20would%20prevent%20a%20silent%20regression%20if%20the%20try%2Fcatch%20is%20ever%20refactored.%0A%0A&repo=ndycode%2Foc-codex-multi-auth&pr=173&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
lib/tools/codex-doctor.ts:155-159
**redundant `changedByRefresh = true` in stale-summary block**

`refreshedAccounts.length > 0` is only true when at least one account passed the `refreshResult.type === "success"` branch, which already sets `changedByRefresh = true` inside the loop. the second assignment here can never flip the flag from `false` to `true` — it's always redundant. a future dev adding a path where stale state is cleared *without* a successful token refresh (e.g. a manual cooldown reset) would silently rely on this flag being set by the refresh loop, creating a subtle dependency that isn't obvious at all.

### Issue 2 of 2
test/stale-state.test.ts:1-10
**missing vitest coverage: `clearTuiQuotaSnapshot` EBUSY failure path**

the catch block at `codex-doctor.ts:200-208` explicitly calls out the windows EBUSY scenario (TUI holds the cache file open), but there's no test that mocks `clearTuiQuotaSnapshot` to throw and asserts the error lands in `fixErrors` rather than crashing the tool. given this codebase ships on windows, a test locking in that error-surface behaviour would prevent a silent regression if the try/catch is ever refactored.

`````

</details>

<sub>Reviews (3): Last reviewed commit: ["fix(codex-doctor): only report stale-sta..."](https://github.com/ndycode/oc-codex-multi-auth/commit/b328500b547beba20134430ae970214e2f81b31d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38599336)</sub>

### CODE REVIEW

File: lib/tools/codex-doctor.ts [hunk 1]
```diff
@@ -130,6 +182,20 @@ export function createCodexDoctorTool(ctx: ToolContext): ToolDefinition {
...
+				try {
+					await clearTuiQuotaSnapshot();
+					appliedFixes.push("Cleared stale TUI quota cache.");
+				} catch (error) {
+					fixErrors.push(
+						`Failed to clear TUI quota cache: ${
+							error instanceof Error ? error.message : String(error)
+						}`,
+					);
+				}
```
[the remark this answers is not in the document]
Comment (ndycode) [reply lib/tools/codex-doctor.ts:209]: Fixed in b328500. `clearTuiQuotaSnapshot` is now guarded behind `changedByRefresh`, so a run where every refresh failed (`refreshedAccounts` empty) no longer clears the cache or reports a phantom "Cleared stale TUI quota cache." Added a catch-block comment noting the Windows `EBUSY` (cache file held open by the TUI) vs `ENOENT` case for future debuggers.

File: test/index.test.ts [hunk 2]
```diff
@@ -1895,6 +1895,61 @@ describe("OpenAIOAuthPlugin", () => {
...
+			const result = await plugin.tool["codex-doctor"].execute({ deep: false });
+
+			expect(result).toContain("disabled duplicate account entry");
+			expect(result).toContain("codex-remove");
+		});
+
+
 		it("returns json output for deep diagnostics", async () => {
 			mockStorage.accounts = [{ refreshToken: "r1", email: "user@example.com" }];
 			const result = parseJsonOutput<{
```
[the remark this answers is not in the document]
Comment (ndycode) [reply test/index.test.ts:1982]: Added in b328500: a vitest case (`does not report cleared stale-state when saveAccounts fails during fix`) mocks `saveAccounts` to reject, then asserts the output contains no `Cleared cooldown` / `stale rate-limit marker` message and does contain `Failed to persist refresh updates`. Combined with the P1 fix (messages moved inside the `saveAccounts` try-block), this locks in the ordering so a failed persist can never be reported as an applied fix.
